sbor/src/traversal/untyped/traverser.rs

```diff
@@ -157,313 +119,393 @@ pub enum ExpectedStart<X: CustomValueKind> {
+                if is_byte_array {
+                    // We know this is >= 1 from the above check
+                    let array_length = container_child_size;
+                    let max_index_which_would_be_read = array_length - 1;
+                    // Set current child index before we read so that if we get an error on read
+                    // then it comes through at the max child index we attempted to read.
+                    parent.current_child_index = max_index_which_would_be_read;
+                    ActionHandler::new_from_current_offset(ancestor_path, decoder)
+                        .read_byte_array(array_length)
+                } else {
+                    // NOTE: parent.current_child_index is already 0, so no need to change it
+                    let implicit_value_kind = parent.get_implicit_value_kind_of_current_child();
+                    ActionHandler::new_from_current_offset(ancestor_path, decoder)
+                        .read_value(implicit_value_kind)
+                }
             }
-            NextEventOverride::None => {
-                let parent = self.container_stack.last();
+            NextAction::ReadNextChildOrExitContainer => {
+                let parent = ancestor_path.last_mut();
                 match parent {
                     Some(parent) => {
-                        if parent.is_complete() {
-                            self.exit_container()
+                        let next_child_index = parent.current_child_index + 1;
+                        let is_complete =
+                            next_child_index >= parent.container_header.get_child_count();
+                        if is_complete {
+                            // We pop the completed parent from the ancestor list
+                            let AncestorState {
+                                container_header,
+                                container_start_offset,
+                                ..
+                            } = ancestor_path.pop().expect("Parent has just been read");
+
+                            ActionHandler::new_with_fixed_offset(
+                                ancestor_path,
+                                decoder,
+                                container_start_offset,
+                            )
+                            .complete_container_end(container_header)
                         } else {
-                            self.read_child_value()
+                            parent.current_child_index = next_child_index;
+                            let implicit_value_kind =
+                                parent.get_implicit_value_kind_of_current_child();
+                            ActionHandler::new_from_current_offset(ancestor_path, decoder)
+                                .read_value(implicit_value_kind)
                         }
                     }
-                    None => self.read_end(),
+                    None => {
```

Comment:
Have added the following comment here on the next PR:
```
                        // We are due to read another element and exit but have no parent
                        // This is because we have finished reading the `root` value.
                        // Therefore we call `end`.
```